src/core/ts/background-scripts/background.js/telemetry/Telemetry.ts

```diff
@@ -5,78 +5,123 @@
 import Glean from "@mozilla/glean/webext";
 import { custom } from "./generated/pings";
 import { config } from "../../../config";
+import {
+  modelLoadTime,
+  translationTime,
+  wordsPerSecond,
+} from "./generated/performance";
+import { fromLang, toLang } from "./generated/metadata";
+import {
+  displayed,
+  changeLang,
+  closed,
+  neverTranslateSite,
+  translate,
+  neverTranslateLang,
+  notNow,
+} from "./generated/infobar";
 
+/**
+ * This class contains general telemetry initialization and helper code and synchronous telemetry-recording functions.
+ *
+ * Synchronous methods here is important, since it is the only way to guarantee that multiple Glean API calls are
+ * executed sequentially and not interleaved with other asynchronous Telemetry recording.
+ * For more information, see: https://github.com/mozilla-extensions/bergamot-browser-extension/pull/76#discussion_r602128568
+ */
 export class Telemetry {
-  private static _instance = null;
-  private _metricsToSubmit;
-
   constructor() {
     const appId = config.telemetryAppId;
     Glean.initialize(appId, true, {
       debug: { logPings: config.telemetryDebugMode },
     });
-    this._metricsToSubmit = 0;
     console.info(
       `Telemetry: initialization completed with application ID ${appId}.`,
     );
-    // Disabled telemetry until opt-out-preference is respected (https://github.com/mozilla-extensions/bergamot-browser-extension/issues/51)
-    this.setUploadEnabled(false);
-    console.info(`Note: Telemetry upload is disabled in this build`);
   }
 
-  /**
-   * Provides access to Telemetry singleton.
-   */
-  public static get global(): Telemetry {
-    if (Telemetry._instance === null) {
-      Telemetry._instance = new Telemetry();
-    }
-    return Telemetry._instance;
+  public onInfoBarDisplayed(tabId: number) {
+    displayed.record();
+    this.submit();
   }
 
-  /**
-   * Enables uploading Glean metrics to telemetry server.
-   * @param val bool
-   */
-  public setUploadEnabled(val: boolean) {
-    Glean.setUploadEnabled(val);
+  public onSelectTranslateFrom(tabId: number) {
+    changeLang.record();
+    this.submit();
+  }
+
+  public onSelectTranslateTo(tabId: number) {
+    changeLang.record();
+    this.submit();
+  }
+
+  public onInfoBarClosed(tabId: number) {
+    closed.record();
+    this.submit();
```

Comment:
I agree, but for this we first need to define what a translation session is. Is it for instance "on a tab-specific basis, all time from browser start until a specific tab has been translated"? Or is it a but broader, including several translations across several tabs? As discussed with @brizental above, it may make sense to use different custom pings here - one for each tab-translation and one for all general user interactions. Let's discuss in a separate issue.

Reply:
@eu9ene since this implementation will change before we will start broad testing, and traffic volume is thus expected to be low, are you ok with keeping it as is in this PR?

Reply:
> Let's discuss in a separate issue.

Filed issue for this: https://github.com/mozilla-extensions/bergamot-browser-extension/issues/94

Reply:
I'm ok with doing things incrementally, but it was a complete deal-breaker in my PR, so I expected this to be preserved here too. It's up to you when to do it, but this thing should be prioritized because it also affects analytics. I believe issue #94 doesn't reflect the problem here - it is not related to how we define user metrics it just how we technically group things together to reduce the number of sent pings and the tab translation session seems like a good candidate to me. It should be 1, max 2 pings per regular page translation.

Reply:
Let's leave it up to @brizental to decide if it is a dealbreaker at this stage. I prefer to fix it properly, eg using separate custom pings, rather than create an interim solution in this PR.

Reply:
@motin ping splitting is up to us and depends on our analytics needs. It's a good suggestion to think more about it, but I assume it wouldn't change until we figure out the complete experimentation plan on real users which might take a while. Glean telemetry is designed to bundle things together and it is also might be more convenient to analyze when it's in one row, so yes, this is something we need to do for sure. The question is when. Basically, it was the main struggle in my PR and caused issue #49 and I expected that we deal with it in Milestone 2.

Reply:
To keep things incremental, I vouch for adding a test that verifies that #49 is solved but we revisit ping grouping in a follow-up PR.

Reply:
Sure, I'm ok with merging faster and creating a new PR, let's just have a note somewhere about it to not forget